docs/configuration.md
# Configuration

___

```python
from furbain import config
```
___

## Introduction

Once you have opened your IDE, you can create a new python file and write `import furbain` and execute it. This will generate the configuration file.  

On windows, furbain's configuration file is located at `C:\Users\name\.furbain\config.json`.  
This file contains the database credentials and the path to the MATSim output files.  

To modifiy these informations you can either edit the file manually or use the following functions:

## Functions

### User

`config.setDatabaseUser(user)` : sets the database user. (string default: `postgres`)  
`config.getDatabaseUser()` : returns the database user.  


### Password

`config.setDatabasePassword(password)` : sets the database password. (string default: `postgres`)  
`config.getDatabasePassword()` : returns the database password.  


### Host

`config.setDatabaseHost(host)` : sets the database host. (string default: `localhost`)  
`config.getDatabaseHost()` : returns the database host.  

### Port

`config.setDatabasePort(port)` : sets the database port. (string default: `5432`)  
`config.getDatabasePort()` : returns the database port.  

### SRID

`config.setDatabaseSRID(srid)` : sets the database SRID. (string default: `4326`)  
`config.getDatabaseSRID()` : returns the database SRID.  

### Simulation output path

`config.setSimulationOutputPath(path)` : sets the path to the MATSim output files. (string default: `C:/Users/name/Documents/matsimOutput/simulation_output`)  
**On windows, do not use a backslash `\` but simple slash `/` in the path**  
`getSimulationOutputPath()` : returns the path to the MATSim output files.  

## Set or get a variable in the configuration file

`config.setVariableInConfigurationFile(name, value)` : sets a variable in the configuration file.  
`config.getVariableInConfigurationFile(name)` : returns a variable in the configuration file.  

Here are the variables you can set with their default values :

```json
{
    "db_host": "localhost",
    "db_port": "5432",
    "db_user": "postgres",
    "db_password": "postgres",
    "db_srid": "2154",
    "path_simulation_output": "C:/Users/name/Documents/matsimOutput/simulation_output",
    "allvehicles_filename": "output_allvehicles.xml.gz",
    "events_filename": "output_events.xml.gz",
    "facilities_filename": "output_facilities.xml.gz",
    "households_filename": "output_households.xml.gz",
    "legs_filename": "output_legs.csv.gz",
    "network_filename": "output_network.xml.gz",
    "persons_filename": "output_persons.csv.gz",
    "plans_filename": "output_plans.xml.gz",
    "experienced_plans_filename": "output_experienced_plans.xml.gz",
    "trips_filename": "output_trips.csv.gz",
    "detailed_network_filename": "detailed_network.csv",
    "buildings_filename": "BUILDINGS.geojson"
}
```

Below `path_simulation_output` are the filenames of the output files of your simulation.


docs/getting-started/environment-setup.md
## Setup the environment

*** For this part, I am using Anaconda ***
1. Install `Anaconda` on your machine. You can follow the instructions in this [guide](https://docs.conda.io/projects/conda/en/latest/user-guide/install/index.html). Direct download [here](https://www.anaconda.com/products/distribution).

2. Open Anaconda Navigator and navigate to the "Environments" tab in the left panel.

3. Click on the "Import" button and select the matsimConverterEnv.yaml file located in the ./resources/setup/ folder of the cloned repository.

4. Give a name to your new environment and wait for the setup process to complete.

5. Once the environment is created, **select it** and launch an IDE like VS Code from Anaconda Navigator.

6. Open a terminal in the IDE and run the command `pip install ./resources/setup/matsim_tools-1.0.5-py3-none-any.whl` to install the required matsim_tools package. **Please verify that you are installing the lastest version available in ./resources/setup/**

src/config.py
import sqlalchemy.types as types
import pathlib
import json
import argparse
from os.path import isdir

DB_DBNAME = ''

# Names of the tables in the database
DB_ALLVEHICLES_TABLE = 'vehicle' 
DB_ALLVEHICLES_TYPES_TABLE = 'vehicleType'
DB_EVENTS_TABLE = 'networkLinkTraffic'
DB_FACILITIES_TABLE = 'facility'
DB_HOUSEHOLDS_TABLE = 'household'
DB_NETWORK_TABLE = 'networkLink'
DB_PERSONS_TABLE = 'person'
DB_PLANS_TABLE = 'activity'
DB_TRIPS_TABLE = 'trip'
DB_BUILDINGS_TABLE = 'building'

# Separators for the csv files
PERSONS_CSV_SEPARATOR = ';'
TRIPS_CSV_SEPARATOR = ';'
DETAILED_NETWORK_CSV_SEPARATOR = ','


# ===== QUERIES =====
ARABESQUE_DEFAULT_SRID = '4326' # EPSG used by Arabesque
ARABESQUE_GENERATED_FILES_DIRECTORY_PATH = './output/'

ACTIVITY_SEQUENCES_TABLE_NAME = 'activitySequences'
ACTIVITY_SEQUENCES_TABLE_TIME_FORMAT = '%d_%m_%Y_%H_%M_%S' # time used to indicate when the table was generated, it will be added to the table name
ACTIVITY_SEQUENCES_TABLE_COLUMNS = {
    "id": types.Integer,
    "personId": types.Integer,
    "periodStart": types.Interval,
    "periodEnd": types.Interval,
    "mainActivityId": types.Integer,
    "startActivityId": types.Integer,
    "endActivityId": types.Integer,
    "mainActivityStartTime": types.Interval,
    "mainActivityEndTime": types.Interval,
    "timeSpentInMainActivity": types.Interval
}
ACTIVITY_SEQUENCES_DB_PROGRESS_BAR_PERCENTAGE = 5 # the progress bar will be updated every 10% of the table


# ===== CONFIGURATION ENV =====
PATH_CONFIGURATION_FILE = pathlib.Path.home() / '.furbain' / 'config.json'


def createConfigurationFile():
    # Create the config file if it doesn't exist
    fileToCreate = PATH_CONFIGURATION_FILE
    
    if not fileToCreate.exists():
        fileToCreate.parent.mkdir(parents=True, exist_ok=True)
        fileToCreate.touch()
        fileToCreate.write_text("""{
    "db_host": "localhost",
    "db_port": "5432",
    "db_user": "postgres",
    "db_password": "postgres",
    "db_srid": "2154",
    "path_simulation_output": "C:/Users/name/Documents/matsimOutput/simulation_output",
    "allvehicles_filename": "output_allvehicles.xml.gz",
    "events_filename": "output_events.xml.gz",
    "facilities_filename": "output_facilities.xml.gz",
    "households_filename": "output_households.xml.gz",
    "legs_filename": "output_legs.csv.gz",
    "network_filename": "output_network.xml.gz",
    "persons_filename": "output_persons.csv.gz",
    "plans_filename": "output_plans.xml.gz",
    "experienced_plans_filename": "output_experienced_plans.xml.gz",
    "trips_filename": "output_trips.csv.gz",
    "detailed_network_filename": "detailed_network.csv",
    "buildings_filename": "BUILDINGS.geojson"
}""")

def loadConfigurationFile():
    fileToLoad = PATH_CONFIGURATION_FILE

    if not fileToLoad.exists():
        createConfigurationFile()
    
    with open(fileToLoad) as json_file:
        return json.load(json_file)


def saveConfigurationFile(config):
    fileToSave = PATH_CONFIGURATION_FILE

    if not fileToSave.exists():
        createConfigurationFile()
    
    with open(fileToSave, 'w') as outfile:
        json.dump(config, outfile, indent=4)


def setVariableInConfigurationFile(name, value):
    config = loadConfigurationFile()
    config[name] = value
    saveConfigurationFile(config)

def getVariableInConfigurationFile(name):
    config = loadConfigurationFile()
    return config[name]

# ----- User -----
def setDatabaseUser(user):    
    setVariableInConfigurationFile('db_user', user)

def getDatabaseUser():
    return getVariableInConfigurationFile('db_user')

def cmdSetDatabaseUser(args=None):
    parser = argparse.ArgumentParser(description='Set the database user')
    parser.add_argument('user', help='The user to use to connect to the database')
    args = parser.parse_args(args)
    setDatabaseUser(args.user)
    print('The database user has been set')


# ----- Password -----
def setDatabasePassword(password):
    setVariableInConfigurationFile('db_password', password)

def getDatabasePassword():
    return getVariableInConfigurationFile('db_password')


# ----- Host -----
def setDatabaseHost(host):
    setVariableInConfigurationFile('db_host', host)

def getDatabaseHost():
    return getVariableInConfigurationFile('db_host')


# ----- Port -----
def setDatabasePort(port):
    setVariableInConfigurationFile('db_port', port)

def getDatabasePort():
    return getVariableInConfigurationFile('db_port')


# ----- Database SRID -----
def setDatabaseSRID(srid):
    setVariableInConfigurationFile('db_srid', srid)

def getDatabaseSRID():
    return getVariableInConfigurationFile('db_srid')


# ----- Simulation output paths -----
def setSimulationOutputPath(path):
    # Add '/' at the end if it doesn't exist
    if path[-1] != '/':
        path += '/'
    
    # Check if the path exists
    if not isdir(path):
        raise Exception('The path ' + path + ' doesn\'t exist')
    
    setVariableInConfigurationFile('path_simulation_output', path)

def getSimulationOutputPath():
    return getVariableInConfigurationFile('path_simulation_output')


# ----- Output files paths -----
def getAllVehiclesPath():
    return getSimulationOutputPath() + getVariableInConfigurationFile('allvehicles_filename')

def getEventsPath():
    return getSimulationOutputPath() + getVariableInConfigurationFile('events_filename')

def getFacilitiesPath():
    return getSimulationOutputPath() + getVariableInConfigurationFile('facilities_filename')

def getHouseholdsPath():
    return getSimulationOutputPath() + getVariableInConfigurationFile('households_filename')

def getLegsPath():
    return getSimulationOutputPath() + getVariableInConfigurationFile('legs_filename')

def getNetworkPath():
    return getSimulationOutputPath() + getVariableInConfigurationFile('network_filename')

def getPersonsPath():
    return getSimulationOutputPath() + getVariableInConfigurationFile('persons_filename')

def getPlansPath():
    return getSimulationOutputPath() + getVariableInConfigurationFile('plans_filename')

def getExperiencedPlansPath():
    return getSimulationOutputPath() + getVariableInConfigurationFile('experienced_plans_filename')

def getTripsPath():
    return getSimulationOutputPath() + getVariableInConfigurationFile('trips_filename')

def getDetailedNetworkPath():
    return getSimulationOutputPath() + getVariableInConfigurationFile('detailed_network_filename')

def getBuildingsPath():
    return getSimulationOutputPath() + getVariableInConfigurationFile('buildings_filename')